packages/flutter_tools/lib/src/runner/flutter_command_runner.dart

```diff
@@ -390,4 +390,33 @@ class FlutterCommandRunner extends CommandRunner {
     if (ArtifactStore.flutterRoot == null)
       ArtifactStore.flutterRoot = _defaultFlutterRoot;
   }
+
+  /// Get all pub packages in the Flutter repo.
+  List<Directory> getRepoPackages() {
+    String root = path.absolute(ArtifactStore.flutterRoot);
+    List<String> directories = <String>[];
+
+    // dev/ (excluding remove dev/docs)
+    directories.addAll(_gatherProjectPaths(path.join(root, 'dev')));
+    directories.removeWhere((String dir) => dir.endsWith('docs'));
+
+    // examples/
+    directories.addAll(_gatherProjectPaths(path.join(root, 'examples')));
+
+    // packages/
+    directories.addAll(_gatherProjectPaths(path.join(root, 'packages')));
+
+    return directories.map((String dir) => new Directory(dir)).toList();
+  }
+
+  static List<String> _gatherProjectPaths(String rootPath) {
+    if (FileSystemEntity.isFileSync(path.join(rootPath, 'pubspec.yaml')))
+      return <String>[rootPath];
+
+    return new Directory(rootPath)
+      .listSync(followLinks: false)
+      .expand((FileSystemEntity entity) {
+        return entity is Directory ? _gatherProjectPaths(entity.path) : <String>[];
+      });
+  }
```

Comment:
Honestly at this point why not just walk the entire repo, rather than just dev, examples, and packages. If we're worried about finding pubspec.yamls in bin/cache, we could be _really_ clever and at each level skip any directories mentioned in .gitignore files. Or, skip any directories that contain a magic sentinel like ".do_not_analyze", and put that in bin/cache alongside the other magical files we have there.